File: 5-Problem-A-Day/46. Permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new LinkedList<>();
        List<Integer> numsList = new ArrayList<>();
        for (int num : nums)
            numsList.add(num);
        backtrack(nums.length, numsList, ans, 0);
        return ans;
    }

    private void backtrack(int n, List<Integer> nums, List<List<Integer>> ans, int first) {
        if (first == n)
            ans.add(new ArrayList<>(nums));
        else {
            for (int i = first; i < n; i++) {
                Collections.swap(nums, first, i);
                backtrack(n, nums, ans, first + 1);
                Collections.swap(nums, first, i);
            }
        }
    }
}

File: 2023/46. Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        helper(nums, 0, res);
        return res;
    }

    void helper(vector<int> nums, int index, vector<vector<int>>& res) {
        if (index == nums.size() - 1) {
            res.push_back(nums);
            return;
        }
        for (int i = index; i < nums.size(); i++) {
            swap(nums[index], nums[i]);
            helper(nums, index + 1, res);
        }
    }
};